File: LTE_record_process/+ltepipe/Runtime.m
classdef Runtime < handle
%RUNTIME Shared initialization context and infrastructure services.

    properties (Access = private)
        Context
        Services
    end

    methods
        function obj = Runtime()
            obj.Context = struct();
            obj.Services = struct();
        end

        function setContext(obj, key, value)
            key = obj.validateKey(key);
            obj.Context.(key) = value;
        end

        function value = getContext(obj, key)
            key = obj.validateKey(key);
            if ~isfield(obj.Context, key)
                error('ltepipe:Runtime:MissingContext', ...
                    'Runtime context "%s" is not available.', key);
            end
            value = obj.Context.(key);
        end

        function addService(obj, key, value)
            key = obj.validateKey(key);
            if isfield(obj.Services, key)
                error('ltepipe:Runtime:DuplicateService', ...
                    'Runtime service "%s" is already registered.', key);
            end
            obj.Services.(key) = value;
        end

        function value = getService(obj, key)
            key = obj.validateKey(key);
            if ~isfield(obj.Services, key)
                error('ltepipe:Runtime:MissingService', ...
                    'Runtime service "%s" is not available.', key);
            end
            value = obj.Services.(key);
        end

        function status = getStatus(obj)
            status = struct('ContextKeys', {fieldnames(obj.Context).'}, ...
                'Services', struct());
            serviceKeys = fieldnames(obj.Services);
            for index = 1:numel(serviceKeys)
                key = serviceKeys{index};
                service = obj.Services.(key);
                if ismethod(service, 'getStatus')
                    status.Services.(key) = service.getStatus();
                else
                    status.Services.(key) = struct( ...
                        'Class', class(service));
                end
            end
        end
    end

    methods (Static, Access = private)
        function key = validateKey(key)
            key = char(string(key));
            if ~isvarname(key)
                error('ltepipe:Runtime:InvalidKey', ...
                    'Runtime keys must be valid MATLAB identifiers.');
            end
        end
    end
end

File: LTESen_python/ltesen/ltepipe/runtime.py
"""Shared initialization context and services."""

from __future__ import annotations

from typing import Any


class Runtime:
    def __init__(self) -> None:
        self._context: dict[str, Any] = {}
        self._services: dict[str, Any] = {}

    @staticmethod
    def _validate_key(key: str) -> str:
        if not isinstance(key, str) or not key.isidentifier():
            raise ValueError("Runtime keys must be valid Python identifiers")
        return key

    def set_context(self, key: str, value: Any) -> None:
        self._context[self._validate_key(key)] = value

    def get_context(self, key: str) -> Any:
        key = self._validate_key(key)
        if key not in self._context:
            raise KeyError(f'Runtime context "{key}" is not available')
        return self._context[key]

    def add_service(self, key: str, value: Any) -> None:
        key = self._validate_key(key)
        if key in self._services:
            raise ValueError(f'Runtime service "{key}" is already registered')
        self._services[key] = value

    def get_service(self, key: str) -> Any:
        key = self._validate_key(key)
        if key not in self._services:
            raise KeyError(f'Runtime service "{key}" is not available')
        return self._services[key]

    def get_status(self) -> dict[str, Any]:
        services: dict[str, Any] = {}
        for key, service in self._services.items():
            get_status = getattr(service, "get_status", None)
            services[key] = (
                get_status() if callable(get_status) else {"class": type(service).__name__}
            )
        return {"context_keys": sorted(self._context), "services": services}


__all__ = ["Runtime"]
